File: tests/Generics.pas
namespace Demo;

interface

uses System.Collections.Generic, Newtonsoft.Json.Linq, System.Collections, System.IO;

type
  MyList = public class(List<String>)
  public
    method Foo;
  end;

  GenExample = public class
  public
    method UseList(l: List<String>);
    method Parse(ret: JObject);
    method WithNested(d: Dictionary<String, List<FileInfo>>);
  end;

  GenericStatic = public class
  public
    class method Use();
  end;

  GenClass = public class
  public
    class procedure DeSerialize<T>(json: String; var obj: T);
  end;

implementation

method MyList.Foo;
begin
end;

method GenExample.UseList(l: List<String>);
var
  v: List<Integer>;
begin
  v := new List<Integer>;
  l.AddRange(v);
end;

method GenExample.Parse(ret: JObject);
var
  aux: Hashtable;
begin
  aux := (ret as JObject).ToObject<Hashtable>();
end;

method GenExample.WithNested(d: Dictionary<String, List<FileInfo>>);
var
  x: Dictionary<String, List<FileInfo>>;
begin
  x := new Dictionary<String, List<FileInfo>>;
end;

class method GenericStatic.Use();
begin
  System.Collections.Generic.List<String>.Sort(nil);
end;

class procedure GenClass.DeSerialize<T>(json: String; var obj: T);
begin
end;

end.

File: tests/Generics.cs
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using System.Collections;
using System.IO;

namespace Demo {
    public partial class MyList : List<string> {
        public void Foo() {
        }
    }
    
    public partial class GenExample {
        public void UseList(List<string> l) {
            List<int> v;
            v = new List<int>();
            l.AddRange(v);
        }
        public void Parse(JObject ret) {
            Hashtable aux;
            aux = (ret as JObject).ToObject<Hashtable>();
        }
        public void WithNested(Dictionary<string, List<FileInfo>> d) {
            Dictionary<string, List<FileInfo>> x;
            x = new Dictionary<string, List<FileInfo>>();
        }
    }
    
    public partial class GenericStatic {
        public static void Use() {
            System.Collections.Generic.List<string>.Sort(null);
        }
    }
    
    public partial class GenClass {
        public static void DeSerialize<T>(string json, T obj) {
        }
    }
}
